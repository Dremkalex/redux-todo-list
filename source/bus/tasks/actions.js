// Types
import { types } from './types';

export const tasksActions = {
    // Sync
    fillTasks: (tasks) => ({
        type:    types.FILL_TASKS,
        payload: tasks,
    }),

    createTask: (task) => ({
        type:    types.CREATE_TASK,
        payload: task,
    }),

    removeTask: (taskId) => ({
        type:    types.REMOVE_TASK,
        payload: taskId,
    }),

    updateTask: (updatedTask) => ({
        type:    types.UPDATE_TASK,
        payload: updatedTask,
    }),

    updateNewTaskMessage: (message) => ({
        type:    types.UPDATE_NEW_TASK_MESSAGE,
        payload: message,
    }),

    updateEditTaskMessage: (message) => ({
        type:    types.UPDATE_EDIT_TASK_MESSAGE,
        payload: message,
    }),

    updateTasksFilter: (message) => ({
        type:    types.UPDATE_TASKS_FILTER,
        payload: message,
    }),

    setEditTaskId: (editTaskId) => ({
        type:    types.SET_EDIT_TASK_ID,
        payload: editTaskId,
    }),

    resetEditTaskId: () => ({
        type: types.RESET_EDIT_TASK_ID,
    }),

    // Async
    fetchTasksAsync: () => ({
        type: types.FETCH_TASKS_ASYNC,
    }),

    createTaskAsync: (newTaskMessage) => ({
        type:    types.CREATE_TASK_ASYNC,
        payload: newTaskMessage,
    }),

    removeTaskAsync: (taskId) => ({
        type:    types.REMOVE_TASK_ASYNC,
        payload: taskId,
    }),

    updateTaskAsync: (updatedTask) => ({
        type:    types.UPDATE_TASK_ASYNC,
        payload: updatedTask,
    }),

    completeAllTasksAsync: (tasks) => ({
        type:    types.COMPLETE_ALL_TASK_ASYNC,
        payload: tasks,
    }),
};
