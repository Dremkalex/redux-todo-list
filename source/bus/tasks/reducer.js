// Core
import { fromJS, Map, List } from 'immutable';

// Types
import { types } from './types';

const initialState = Map({
    tasks:           List(),
    newTaskMessage:  '',
    editTaskMessage: '',
    editTaskId:      '',
    tasksFilter:     '',
});

export const tasksReducer = (state = initialState, { type, payload }) => {
    switch (type) {
        case types.FILL_TASKS:
            return state.set('tasks', fromJS(payload));

        case types.CREATE_TASK:
            return state.set('tasks', state.get('tasks').unshift(fromJS(payload)));

        case types.REMOVE_TASK:
            return state.set('tasks',
                state.get('tasks').
                    filter((task) => task.get('id')!== payload));

        case types.UPDATE_TASK:
            return state.set('tasks',
                state.get('tasks')
                    .update(state.get('tasks').findIndex((task) => task.get('id') === payload.id),
                        (task) => task.merge(payload)));

        case types.UPDATE_NEW_TASK_MESSAGE:
            return state.set('newTaskMessage', payload);

        case types.UPDATE_EDIT_TASK_MESSAGE:
            return state.set('editTaskMessage', payload);

        case types.SET_EDIT_TASK_ID:
            return state.set('editTaskId', payload);

        case types.RESET_EDIT_TASK_ID:
            return state.set('editTaskId', '');

        case types.UPDATE_TASKS_FILTER:
            return state.set('tasksFilter', payload);

        default:
            return state;
    }
};
