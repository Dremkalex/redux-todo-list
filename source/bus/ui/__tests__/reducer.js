// Reducer
import { uiReducer } from "../reducer";

// Actions
import { uiActions } from "../actions";

describe("ui reducer:", () => {
    test("should return initial state by default", () => {
        expect(uiReducer(void 0, {})).toMatchSnapshot();
    });

    test("should handle a START_FETCHING action", () => {
        expect(uiReducer(void 0, uiActions.startFetching())).toMatchSnapshot();
    });

    test("should handle a STOP_FETCHING action", () => {
        expect(
            uiReducer(void 0, uiActions.stopFetching())).toMatchSnapshot();
    });
});
