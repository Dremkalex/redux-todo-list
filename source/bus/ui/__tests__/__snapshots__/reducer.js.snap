// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui reducer: should handle a START_FETCHING action 1`] = `
Immutable.Map {
  "isTasksFetching": true,
}
`;

exports[`ui reducer: should handle a STOP_FETCHING action 1`] = `
Immutable.Map {
  "isTasksFetching": false,
}
`;

exports[`ui reducer: should return initial state by default 1`] = `
Immutable.Map {
  "isTasksFetching": false,
}
`;
